File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-7.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        while l <= r:
            middle_idx = l + int((r - l) / 2)
            if nums[middle_idx] == target: return middle_idx
            elif nums[l] == target: return l
            elif nums[r] == target: return r
            if nums[l] <= nums[middle_idx]:
                if target < nums[middle_idx] and target > nums[l]: r = middle_idx - 1
                else: l = middle_idx + 1
            else:
                if target > nums[middle_idx] and target < nums[r]: l = middle_idx + 1
                else: r = middle_idx - 1

        return -1

File: Data Structures & Algorithms/search-insert-position/submission-7.cpp
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        if (nums.size() < 2) return target <= nums[0] ? 0 : 1;
        int max_elem = *max_element(nums.begin(), nums.end());
        if (max_elem < target) return nums.size();
        int l = 0; int r = nums.size() - 1;

        while (l <= r) {
            int midIdx = (l + r) / 2;
            if (nums[l] == target) return l;
            else if (nums[r] == target) return r;

            if (r - l == 1) {
                if (nums[l] > target) return l;
                return r;
            }
            else if (r - midIdx == 1 && midIdx - l == 1) {
                if (target > nums[l] && target < nums[midIdx]) return midIdx;
                else if (target > nums[midIdx] && target < nums[r]) return r;
                else if (target > nums[r]) return r;
            }

            if (nums[midIdx] < target) l = midIdx + 1;
            else if (nums[midIdx] > target) r = midIdx;
            else return midIdx;
        }

        return r;
    }
};